pyoozie/builder.py

```diff
@@ -27,56 +28,77 @@ def _coordinator_submission_xml(username, coord_xml_path, configuration=None, in
 
 class WorkflowBuilder(object):
 
-    def __init__(self, name):
-        # Initially, let's just use a static template and only one action payload and one action on error
-        self._name = tags.validate_xml_name(name)
-        self._action_name = None
-        self._action_payload = None
-        self._action_error = None
-        self._kill_message = None
-
-    def add_action(self, name, action, action_on_error, kill_on_error='${wf:lastErrorNode()} - ${wf:id()}'):
-        # Today you can't rename your action and you can only have one, but in the future you can add multiple
-        # named actions
-        if any((self._action_name, self._action_payload, self._action_error, self._kill_message)):
-            raise NotImplementedError("Can only add one action in this version")
-        else:
-            self._action_name = tags.validate_xml_id('action-' + name)
-            self._action_payload = action
-            self._action_error = action_on_error
-            self._kill_message = kill_on_error
+    def __init__(self, name, action_layout_strategy=transforms.serial_layout, parameters=None,
+                 configuration=None, credentials=None, job_tracker=None, name_node=None, job_xml_files=None,
+                 default_retry_max=None, default_retry_interval=None, default_retry_policy=None):
+        self.__workflow = tags.Workflow(
+            name=name,
+            parameters=parameters,
+            configuration=configuration,
+            credentials=credentials,
+            job_tracker=job_tracker,
+            name_node=name_node,
+            job_xml_files=job_xml_files,
+            default_retry_max=default_retry_max,
+            default_retry_interval=default_retry_interval,
+            default_retry_policy=default_retry_policy,
+        )
+        self.__action_layout_strategy = action_layout_strategy
+        self.__actions = {}
+        self.__action_kwargs = {}
+        self.__dependencies = {}
+        self.__transformations = []
+        self.__transformation_kwargs = {}
+
+    def add_action(self, name, action, depends_upon=None, **kwargs):
+        def create_action_name(name):
+            return tags.validate_xml_id('action-' + name)
+
+        # Validate name
+        name = create_action_name(name)
+        assert name not in self.__actions, "Cannot add an action with the same name (%(name)s) twice".format(
+            name=name)
+
+        # Compose and validate dependency names
+        dependencies = set()
+        for dependency_name in depends_upon or {}:
+            dependencies.add(create_action_name(dependency_name))
+        self.__dependencies[name] = dependencies
+
+        # Store actions
+        self.__actions[name] = action
+        self.__action_kwargs[name] = kwargs
 
         return self
 
+    def add_transform(self, fnc_transform, **kwargs):
```

Comment:
Add a transformation function to be run on the graph of nodes within the workflow and any arguments that it should take.